### PR TITLE
Feature soft autoassignment

### PR DESCRIPTION
**Pull request by bogdan-bondar**
Auto-assignment now supports distribution set assignment with soft semantic.
 
* Action type field in target filter query
* Auto-assignment repository management interface for action type selection
* Optional payload field _type_ for `/autoAssignDS` Mgmt API target filter endpoint 
* Validation of distribution set (completed, not soft deleted) before auto-assignment
* Group of radio buttons to select between _forced_ and _soft_ action type in Auto assignment pop-up window UI
* Combobox with filter function for distribution set selection instead of Table in Auto assignment pop-up window UI
* REST documentation for auto-assignment type
* Hawkbit documentation with auto-assignment description under Target Filter UI
* Debouncing for metadata pop-up window input fields

fixes https://github.com/eclipse/hawkbit/issues/716

### CODE REVIEW

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-core/src/main/java/org/eclipse/hawkbit/exception/SpServerError.java [hunk 1]
```diff
@@ -173,22 +199,38 @@
...
 
     /**
      * Error message informing that the maintenance schedule is invalid.
      */
-    SP_MAINTENANCE_SCHEDULE_INVALID("hawkbit.server.error.maintenanceScheduleInvalid", "Information for schedule, duration or timezone is missing; or there is no valid maintenance window available in future.");
+    SP_MAINTENANCE_SCHEDULE_INVALID("hawkbit.server.error.maintenanceScheduleInvalid",
+            "Information for schedule, duration or timezone is missing; or there is no valid maintenance window available in future."),
+
+    /**
+    *
```
Comment (bogdan-bondar) [inline hawkbit-core/src/main/java/org/eclipse/hawkbit/exception/SpServerError.java]: complete

File: hawkbit-core/src/main/java/org/eclipse/hawkbit/exception/SpServerError.java [hunk 2]
```diff
@@ -173,22 +199,38 @@
...
+    */
+    SP_AUTO_ASSIGN_ACTION_TYPE_INVALID("hawkbit.server.error.repo.invalidAutoAssignActionType",
+            "The given action type for auto-assignment is invalid: allowed values are FORCED and SOFT"),
+
+    /**
+    *
```
Comment (bogdan-bondar) [inline hawkbit-core/src/main/java/org/eclipse/hawkbit/exception/SpServerError.java]: complete

File: hawkbit-core/src/main/java/org/eclipse/hawkbit/repository/TargetFilterQueryFields.java [hunk 3]
```diff
@@ -29,6 +29,11 @@
      */
     NAME("name"),
 
+    /**
+     * action type set for auto assign distribution set
```
Comment (bogdan-bondar) [inline hawkbit-core/src/main/java/org/eclipse/hawkbit/repository/TargetFilterQueryFields.java]: dot a the end

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetFilterQueryManagement.java [hunk 4]
```diff
@@ -230,8 +234,46 @@
...
+    /**
+     * Updates the the auto-assign {@link DistributionSet} and
+     * {@link ActionType} of the addressed {@link TargetFilterQuery}
+     *
+     * @param queryId
+     *            of the target filter query to be updated
+     * @param dsId
+     *            to be updated or <code>null</code> in order to remove it
+     *            together with the auto-assign {@link ActionType}
+     * @param actionType
```
Comment (bogdan-bondar) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetFilterQueryManagement.java]: of null for default (FORCED)

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/JpaTargetFilterQuery.java [hunk 5]
```diff
@@ -57,6 +59,10 @@
     @JoinColumn(name = "auto_assign_distribution_set", nullable = true, foreignKey = @ForeignKey(value = ConstraintMode.CONSTRAINT, name = "fk_filter_auto_assign_ds"))
     private JpaDistributionSet autoAssignDistributionSet;
 
+    @Column(name = "auto_assign_action_type", nullable = true)
+    @Convert("actionType")
```
Comment (bogdan-bondar) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/JpaTargetFilterQuery.java]: define converter with different name explicitly

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/dstable/ManageDistBeanQuery.java [hunk 6]
```diff
@@ -72,13 +85,24 @@ public ManageDistBeanQuery(final QueryDefinition definition, final Map<String, O
                 dsComplete = (Boolean) queryConfig.get(SPUIDefinitions.FILTER_BY_DS_COMPLETE);
             }
         }
+    }
+
+    private void setFilterString(final QueryDefinition definition) {
+        // if search text is set, we do not want to apply the filter
+        if (StringUtils.isEmpty(searchText)) {
+            filterString = definition.getFilters().stream().filter(SimpleStringFilter.class::isInstance)
+                    .map(SimpleStringFilter.class::cast).map(SimpleStringFilter::getFilterString).findFirst()
```
Comment (bogdan-bondar) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/dstable/ManageDistBeanQuery.java]: findAny

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/filtermanagement/DistributionSetSelectWindow.java [hunk 7]
```diff
@@ -128,25 +120,36 @@ public void showForTargetFilter(final Long tfqId) {
...
+        actionTypeOptionGroupLayout.getActionTypeOptionGroup().select(actionTypeToSet);
+
+        if (distributionSet != null) {
+            final String initialFilterString = HawkbitCommonUtil.getFormattedNameVersion(distributionSet.getName(),
+                    distributionSet.getVersion());
+            final int filteredSize = dsCombo.setInitialValueFilter(initialFilterString);
+
+            if (filteredSize <= 0) {
+                notification.displayValidationError(
+                        i18n.getMessage("message.selected.distributionset.not.found", initialFilterString));
```
Comment (bogdan-bondar) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/filtermanagement/DistributionSetSelectWindow.java]: use constant

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDistributionSetManagement.java [hunk 8]
```diff
@@ -642,6 +652,19 @@ public boolean isInUse(final long setId) {
         return specList;
     }
 
+    // the format of filter string is 'name:version'. 'name' and 'version'
+    // fields follow the starts_with semantic, that changes to equal for 'name'
+    // field when the semicolon is present
+    private static String[] getDsFilterNameAndVersionEntries(final String filterString) {
```
Comment (blomark) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDistributionSetManagement.java]: Please extract this method to an utility class and write unit tests

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/TargetFilterQueryManagementTest.java [hunk 9]
```diff
@@ -180,20 +185,65 @@ public void updateTargetFilterQuery() {
...
+            final DistributionSet distributionSet) {
+        // assigning a distribution set with TIMEFORCED action is supposed to
+        // fail as only FORCED and SOFT action types are allowed
+        assertThatExceptionOfType(InvalidAutoAssignActionTypeException.class).isThrownBy(
+                () -> targetFilterQueryManagement.updateAutoAssignDSWithActionType(targetFilterQuery.getId(),
+                        distributionSet.getId(), ActionType.TIMEFORCED));
+    }
+
+    @Step
+    private void verifyAutoAssignmentWithInvalidDistributionSet(final TargetFilterQuery targetFilterQuery) {
```
Comment (blomark) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/TargetFilterQueryManagementTest.java]: Please add test that ensures that a soft deleted assigned DS will fail also

File: hawkbit-rest/hawkbit-mgmt-resource/src/test/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtTargetFilterQueryResourceTest.java [hunk 10]
```diff
@@ -340,33 +350,114 @@ public void updateTargetFilterQueryWithQueryThatExceedsQuota() throws Exception
...
+
+        mvc.perform(post(MgmtRestConstants.TARGET_FILTER_V1_REQUEST_MAPPING + "/" + tfq.getId() + "/autoAssignDS")
+                .content("{\"id\":" + incompleteDistributionSet.getId() + "}").contentType(MediaType.APPLICATION_JSON))
+                .andDo(MockMvcResultPrinter.print()).andExpect(status().isBadRequest())
+                .andExpect(jsonPath(JSON_PATH_EXCEPTION_CLASS,
+                        equalTo(InvalidAutoAssignDistributionSetException.class.getName())))
+                .andExpect(jsonPath(JSON_PATH_ERROR_CODE,
+                        equalTo(SpServerError.SP_AUTO_ASSIGN_DISTRIBUTION_SET_INVALID.getKey())));
+    }
+
```
Comment (blomark) [inline hawkbit-rest/hawkbit-mgmt-resource/src/test/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtTargetFilterQueryResourceTest.java]: PLease add test that ensures that a soft deleted DS could not be used for auto assignment

**Pull request merged**